src/io/delimited.cr

```diff
@@ -125,32 +125,23 @@ class IO::Delimited < IO
         peek.index(first_byte)
       end
 
-    # If we can't find the delimiter's first byte we can just read from peek
-    unless index
-      # If we have more in peek than what we need to read, read all of that
-      if peek.size >= slice.size
-        if slice.size == 1
-          # For a size of 1, this is much faster
-          slice[0] = peek[0]
-        else
-          slice.copy_from(peek[0, slice.size])
-        end
-        @io.skip(slice.size)
-        return slice.size
+    # If the matching byte is past the data we need to read,
+    # or if we have more in peek than what we need to read,
+    # read all of that.
```

Comment:
> If we have an `index` and it is bigger than `slice.size`, *or* if we don't have an `index` and `peek.size` is bigger than `slice.size`, we can have more data available than what we need to be read and we can read all of it.

Alternatively, I think we even merge all branches together:
```cr
# The amount of data available to read is limited by the position of `index` if found, or the entire peek data otherwise
available = index ? peek[0, index] : peek
# If there is enough available data to fill `slice`, we can read all of it at once
if available.size >= slice.size
  if slice.size == 1
    # For a size of 1, this is much faster
    slice[0] = available[0]
  else
    slice.copy_from(available)
  end
  @io.skip(slice.size)
  return slice.size
end
```

Reply:
Good idea! But then after that the compiler can't know `index` is not `nil`...

Reply:
I have an idea: what if we merge this without these style suggestions, I fix the wrong behavior, and then you or someone else can send a follow up PR with a different style?